juju/juju
# Pass volume attachments to PrecheckInstance

## Description of change

We extend the Prechecker (now InstancePrechecker) interface's PrecheckInstance method with an additional parameter, the *existing* volumes that are to be attached to the instance. This allows a provider to, say, identify a conflict between zone placement and the zone in which the volumes are found. PrecheckInstance now takes a parameter struct.

The gce, ec2, and openstack providers have been updated to check for placement/volume availability zone conflicts.

## QA steps

1. juju bootstrap (openstack|aws|google)
2. juju deploy postgresql --storage pgdata=10G
(waiit)
3. juju remove-application postgresql
(wait)
4. juju deploy postgresql pg2 --attach-storage pgdata/0 --to zone=<some-other-zone>
(should fail immediately, rather than being recorded in state and failing at provisioning time)

## Documentation changes

None.

## Bug reference

None.
